### ISSUE TITLE
SiteTree not respecting SiteConfig.CanCreateTopLevelType permssion

### ISSUE THREAD

**Issue by sb-relaxt-at** [silverstripe/silverstripe-cms#1333]
Moving a Page by changing its parent Page in the Settings-Tab does not check whether or not the current User has the appropriate permission (SiteConfig#canCreateTopLevel). This permissions is only respected when Drag&Dropping Pages (LeftAndMain#savetreenode). In SiteTree this permission is only checked in canCreate, but from my point of view it should be also checked when editing an existing Page. Otherwise this settings is useless as someone can create the Page at a sub level and move it upwards later on. 

I am not sure, if there is any reason for this behaviour, but I would suggest to fix this. We would need to add only two small things: 

* SiteTree#validate: If the new parent is root, check the SiteConfig permission
* SiteTree#getSettingsFields: Ensure that the option "at root level" is not available, if there are insufficient rights.

Furthermore it should be discussed if a user without the CanCreateTopLevel permission should be allowed to move a page from the root level to a sub level. I would argue against this, as it seems to be counter-intuitive that a Page can be moved downwards but not back upwards. It also seems desirable to have the ability to limit any structural modification at the root level. In this case a more appropriate name might be CanManageTopLevel.

**Comment by tractorcow** [comment]
I'm not sure that moving a file should invoke canCreate at all, because no page is being created. Should it not check canEdit on that object instead?

**Comment by sb-relaxt-at** [comment]
Well moving a Page to the root level is quite the same as creating a new Page at the root level, isn't it? Nevertheless it might be better to rename this permission (SiteConfig.CanCreateTopLevelType) in the long run. Calling canEdit is not an option, especially as this is about the SiteConfig-Setting. Furthermore calling canCreateTopLevel would be consistent to the implementation in LeftAndMain#savetreenode.

**Comment by tractorcow** [comment]
I think the problem here is that we don't have separate permissions for "can edit a subsite of fields". In this case, the field is ParentID. There is in fact no model to support such a permission in any case. :)

Even if it does involve interacting with the top level of the tree, moving a page still doesn't invoke a page creation, as it's simply a single field change.

What you are asking for is a new "CanMovePagesToRoot" permission on the site tree. You can probably implement this in your own code by using validators, and throw validation errors in case a page is moved to root, but fails a custom check.

Have a look at `SiteTree::validate` method. You can override that, and look at it as a base for implementing this logic in your own custom code. If you like, you can also create a SiteConfig extension to store your new permission.

**Comment by sb-relaxt-at** [comment]
Yes overwriting the validate method is absolutely possible. But imho this is a bug that needs to be fixed in the core. First changing the parent via the TreeDropdownField is possible (as it not checks SiteConfig.CanCreateTopLevel), whereas Drag&Dropping the Page in the site tree is not possible (as it checks SiteConfig.CanCreateTopLevel). This is inconsistent. Secondly if this check is not added, the SiteConfig option would be pointless, as it can be easily circumvented. 

I totally agree that there is no generic way to solve permissions-per-field and this is something that needs to be solved by the developer on a per-project-basis. But this is about a very special field and there is a core permission that needs to be checked.

**Comment by tractorcow** [comment]
If the bug is that it is inconsistent, then we could just make both checks use canEdit. :)

However, rushing into changing anything is likely to have behavioural regressions. I'll admit there is a weakness in the current system, but I don't think that treating all ParentID field changes as a canCreate() is the correct behaviour.

**Comment by maxime-rainville** [comment]
Need to confirm if this is still a problem in SS4.
